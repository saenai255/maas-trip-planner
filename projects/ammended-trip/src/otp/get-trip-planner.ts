import Axios from 'axios';
import { Coordinates } from '../dto/coordinates';
import config from '../config';
import TripPlannerResponse from './dto/trip-planner-response';

async function getTripPlanner(from: Coordinates, to: Coordinates): Promise<TripPlannerResponse> {
    const now = new Date();
    const additionalParams: Record<string, string | number | boolean> = {
        time: now.getHours() + ':' + now.getMinutes(),
        date: now.getDate() + '-' + (now.getMonth() + 1) + '-' + now.getFullYear(),
        mode: 'TRANSIT',
        arriveBy: false,
        wheelchair: false,
        showIntermediateStops: true,
        locale: 'en',
    };

    const params = new URLSearchParams({
        fromPlace: from.join(','),
        toPlace: to.join(','),
        ...additionalParams,
    });

    const result = await Axios.get<TripPlannerResponse>(`${config.otp.url}/otp/routers/default/plan?${params.toString()}`, {
        headers: {
            'x-api-key': config.otp.apiKey,
        },
    });

    return result.data;
}

export { getTripPlanner };
