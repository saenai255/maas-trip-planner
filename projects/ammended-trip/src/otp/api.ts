import { Coordinates } from '../dto/coordinates';
import TripPlannerResponse from './dto/trip-planner-response';
import log from '../logger';
import { getTripPlanner } from './get-trip-planner';

async function fetchPlan(from: Coordinates, to: Coordinates): Promise<TripPlannerResponse> {
    try {
        log.debug('Fetching routes based on given coordinates.', {
            from,
            to,
        });

        const result = await getTripPlanner(from, to);

        log.info('Successfully fetched routes based on given coordinates.', {
            from,
            to,
        });

        return result;
    } catch (e) {
        log.error('Failed to fetch routes', { message: e?.message, stack: e?.stack });
        throw new Error('Failed to fetch routes');
    }
}

const api = {
    tripPlanner: {
        get: fetchPlan,
    },
} as const;

export default api;
