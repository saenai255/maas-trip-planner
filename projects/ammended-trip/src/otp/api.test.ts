import api from './api';
import { getTripPlanner } from './get-trip-planner';

jest.mock('./get-trip-planner', () => ({
    getTripPlanner: jest.fn(),
}));

describe('OpenTripPlanner API', () => {
    describe('tripPlanner', () => {
        const getTripPlannerMock = getTripPlanner as jest.MockedFunction<typeof getTripPlanner>;
        it('should return a TripPlannerResponse', async () => {
            const expectedResponse = {} as any;
            getTripPlannerMock.mockResolvedValueOnce(expectedResponse);

            const response = await api.tripPlanner.get([0, 0], [0, 0]);
            expect(getTripPlannerMock).toHaveBeenCalledTimes(1);
            expect(response).toBe(expectedResponse);
        });

        it('should throw an error if the request fails', async () => {
            getTripPlannerMock.mockRejectedValueOnce(new Error('Request failed'));

            await expect(api.tripPlanner.get([0, 0], [0, 0])).rejects.toThrowErrorMatchingSnapshot();
        });
    });
});
