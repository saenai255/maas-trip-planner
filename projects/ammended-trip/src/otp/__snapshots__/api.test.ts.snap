// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenTripPlanner API tripPlanner should throw an error if the request fails 1`] = `"Failed to fetch routes"`;
